### PR TITLE
Enable branch coverage

### PR DESCRIPTION
**Pull request by maddenp-cu**
**Synopsis**

Enable branch coverage and handle side effects to reestablish 100% coverage.

**Type**

- [x] Tooling (CI, code-quality, packaging, revision-control, etc.)

**Impact**

- [x] This is a non-breaking change (existing functionality continues to work as expected)

**Checklist**

- [x] I have added myself and any co-authors to the PR's _Assignees_ list.
- [x] I have reviewed the documentation and have made any updates necessitated by this change.

### CODE REVIEW

File: src/pyproject.toml [hunk 1]
```diff
@@ -6,6 +6,9 @@ fail_under = 100
 show_missing = true
 skip_covered = true
 
+[tool.coverage.run]
+branch = true
```
Comment (maddenp-cu) [inline src/pyproject.toml]: Enable branch coverage.

File: src/uwtools/api/execute.py [hunk 2]
```diff
@@ -134,17 +133,16 @@ def _get_driver_module_explicit(module: Path) -> ModuleType | None:
     :param module: Name of driver module to load.
     """
     log.debug("Loading module %s", module)
-    if spec := spec_from_file_location(module.name, module):
+    if (spec := spec_from_file_location(module.name, module)) and spec.loader:
```
Comment (maddenp-cu) [inline src/uwtools/api/execute.py]: Combine conditions in a `single` if statement to simplify testing.

File: src/uwtools/drivers/jedi.py [hunk 3]
```diff
@@ -43,18 +43,17 @@ def validate_only(self):
         """
         taskname = self.taskname("validate_only")
         yield taskname
-        a = asset(None, lambda: False)
+        ready = [False]
+        a = asset(None, lambda: ready[0])
```
Comment (maddenp-cu) [inline src/uwtools/drivers/jedi.py]: This uses a pattern for [an in-memory `iotaa` asset](https://github.com/maddenp/iotaa#in-memory-asset): Create a mutable object, return its _content_ as the readiness value, then update that content later.

File: src/uwtools/rocoto.py [hunk 4]
```diff
@@ -365,18 +365,15 @@ def _add_task_dependency_metataskdep(self, e: _Element, config: dict) -> None:
...
+    def _add_task_dependency_sh(self, e: _Element, config: dict, name_attr: str) -> None:
         """
         :param e: The parent element to add the new element to.
         :param config: Configuration data for the tag.
         :param name_attr: XML name attribute for element.
         """
         e = self._add_compound_time_string(e, config[STR.command], STR.sh)
         config[STR.attrs] = config.get(STR.attrs, {})
-        if name_attr:
-            config[STR.attrs][STR.name] = name_attr
```
Comment (maddenp-cu) [inline src/uwtools/rocoto.py]: There's no current use case for `name_attr` not being set, so there's no need for this to be conditional.

File: src/uwtools/tests/api/test_execute.py [hunk 5]
```diff
@@ -54,19 +54,21 @@ def test_execute_fail_stdin_not_ok(kwargs, utc):
     assert str(e.value) == "Set stdin_ok=True to permit read from stdin"
 
 
+@mark.parametrize("graph", [True, False])
```
Comment (maddenp-cu) [inline src/uwtools/tests/api/test_execute.py]: The code-under-test has an `if graph_file:` condition that, previously, was never false, but full branch coverage insists on testing the false case, too. This pattern, which repeats later, supports that.

File: src/uwtools/tests/config/formats/test_base.py [hunk 6]
```diff
@@ -73,10 +73,24 @@ def test_config_base__characterize_values(config):
         5: [{"b": 43}],
         6: "string",
         7: "{% for n in range(3) %}{{ n }}{% endfor %}",
+        8: ["{{ 1 + 1 }}"],
+        9: [42],
```
Comment (maddenp-cu) [inline src/uwtools/tests/config/formats/test_base.py]: Extra test data to exercise the negative side of `if` conditionals in the code.

File: src/uwtools/tests/test_rocoto.py [hunk 7]
```diff
@@ -228,30 +230,40 @@ def test_rocoto__RocotoIterator__query_stmt(self, instance):
             == "select state from jobs where taskname=:taskname and cycle=:cycle order by id desc"
         )
 
-    def test_rocoto__RocotoIterator__report(self, instance, logged):
-        instance._database.touch()
+    @mark.parametrize("create_database", [True, False])
```
Comment (maddenp-cu) [inline src/uwtools/tests/test_rocoto.py]: The code-under-test has an `if` condition that checks for database existence. The to-date tests have always provided a database, so the updated test exercises the case where the database does _not_ exist.

**Review by christinaholtNOAA — CHANGES_REQUESTED** [review]

File: src/uwtools/tests/drivers/test_gsi.py [hunk 8]
```diff
@@ -81,10 +82,12 @@ def test_GSI_filelist(driverobj):
...
-    expected_content = """bar
-baz
-foo"""
-    assert filelist.read_text() == expected_content
+    expected_content = """
+    bar
+    baz
+    foo
+    """
+    assert filelist.read_text() == dedent(expected_content).strip()
```
Comment (maddenp-cu) [inline src/uwtools/tests/drivers/test_gsi.py]: I decided to update this to match the indented style present everywhere (AFAIK) else in our tests.

File: src/uwtools/drivers/gsi.py [hunk 9]
```diff
@@ -59,10 +59,9 @@ def namelist_file(self):
             path=path,
             schema=self.namelist_schema(),
         )
-        if path.is_file():
-            obs_input = Path(self.config["obs_input_file"]).read_text()
-            with path.open(mode="a") as nml:
-                nml.write(obs_input)
+        obs_input = Path(self.config["obs_input_file"]).read_text()
+        with path.open(mode="a") as nml:
+            nml.write(obs_input)
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/drivers/gsi.py]: Was the `if` guard meant to be here?
